### DISCUSSION TITLE
Usage in GNU screen broken?

### DISCUSSION THREAD

**Discussion by lingfish** [Textualize/textual#5032] [Q&A]
Is GNU `screen` supported?  I run a fairly stock config of screen under Alacritty, and this is what I see with the demo:

![image](https://github.com/user-attachments/assets/cdc65bba-ae49-4a8e-82f7-88dfa083c1f9)

If I exit `screen` and re-run, it looks fine.

**Comment by TomJGooding** [comment]
It will be helpful if you run the following command and paste the results:

```
textual diagnose
```

If you don't have the `textual` command on your path, you may have forgotten to install the `textual-dev` package.

**Comment by lingfish** [comment]
Yes, of course, oops.  Here it is inside `screen`:

<details><summary>Inside screen</summary>
<p>
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.79.1 |
| Rich    | 13.8.1 |

## Python

| Name           | Value                                                  |
|----------------|--------------------------------------------------------|
| Version        | 3.11.2                                                 |
| Implementation | CPython                                                |
| Compiler       | GCC 12.2.0                                             |
| Executable     | /srv/virtualenv/multimon-ng-pretty-wU1IQLii/bin/python |

## Operating System

| Name    | Value                                               |
|---------|-----------------------------------------------------|
| System  | Linux                                               |
| Release | 6.1.0-23-amd64                                      |
| Version | #1 SMP PREEMPT_DYNAMIC Debian 6.1.99-1 (2024-07-15) |

## Terminal

| Name                 | Value                     |
|----------------------|---------------------------|
| Terminal Application | *Unknown* (inside screen) |
| TERM                 | screen                    |
| COLORTERM            | truecolor                 |
| FORCE_COLOR          | *Not set*                 |
| NO_COLOR             | *Not set*                 |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=192, height=57 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 192                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 57                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |
</p>
</details> 

And this is outside of `screen`:

<details><summary>alacritty only</summary>
<p>

# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.79.1 |
| Rich    | 13.8.1 |

## Python

| Name           | Value                                                  |
|----------------|--------------------------------------------------------|
| Version        | 3.11.2                                                 |
| Implementation | CPython                                                |
| Compiler       | GCC 12.2.0                                             |
| Executable     | /srv/virtualenv/multimon-ng-pretty-wU1IQLii/bin/python |

## Operating System

| Name    | Value                                               |
|---------|-----------------------------------------------------|
| System  | Linux                                               |
| Release | 6.1.0-23-amd64                                      |
| Version | #1 SMP PREEMPT_DYNAMIC Debian 6.1.99-1 (2024-07-15) |

## Terminal

| Name                 | Value     |
|----------------------|-----------|
| Terminal Application | *Unknown* |
| TERM                 | alacritty |
| COLORTERM            | truecolor |
| FORCE_COLOR          | *Not set* |
| NO_COLOR             | *Not set* |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=192, height=58 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 192                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 58                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |
</p>
</details>

**Reply by TomJGooding** [reply]
Thanks, could I also ask what version of `screen` you're using? Do other terminal applications (e.g. htop) show correctly?

**Reply by lingfish** [reply]
`screen 4.09.00 (GNU) 30-Jan-22`

All other apps work perfectly... `htop`, `nvtop` etc.

**Reply by TomJGooding** [reply]
Could you try updating your `screen` version and see if that helps?

**Reply by lingfish** [reply]
Hrm, I'm on bookworm; there is 4.9.1 I think in testing. Do you know > 4.9.0 could/will be a fix?  Otherwise, I'm hunting backports, or having to build from source...

**Reply by TomJGooding** [reply]
Apparently from `screen v4.4` this sets $TERM to screen.$TERM, so I think upgrading is worth trying for the color support.

EDIT: You could also try some of the troubleshooting steps here: https://wiki.archlinux.org/title/GNU_Screen#Use_256_colors

**Reply by willmcgugan** [reply]
I think the issue is screen setting `TERM=screen`. I'm surprised it doesn't pass through the actual value from the terminal.

We could fix it from the Rich side. And assume "xterm-256color" if TERM is screen. Not sure if that is the right thing to do.

Looks like Tom beat me to it!

**Reply by lingfish** [reply]
> Apparently from `screen v4.4` this sets $TERM to screen.$TERM, so I think upgrading is worth trying for the color support.
> 
> EDIT: You could also try some of the troubleshooting steps here: https://wiki.archlinux.org/title/GNU_Screen#Use_256_colors

I didn't mention either that I've manually set/exported variations on `TERM`, including the dotted form, `screen-256color` etc, with no difference. eg. `screen.alaceitty`.

I did this inside the shell, and using `-T`.

**Reply by lingfish** [reply]
> I think the issue is screen setting `TERM=screen`. 

See my reply above 🙂

> 
> We could fix it from the Rich side. And assume "xterm-256color" if TERM is screen. Not sure if that is the right thing to do.
> 
> Looks like Tom beat me to it!

So you think this is a < 256 colour issue?

**Reply by TomJGooding** [reply]
Hmm, interesting. What does `tput colors` report after setting your TERM to `screen-256color` or `alacritty`?

**Reply by lingfish** [reply]
AFK for a bit, but... Just pointing out I'm on 4.09 (4.9?) and you've mentioned 4.4... so I'm > 4.4...

**Reply by TomJGooding** [reply]
Sorry yes I must have misread that - blame me still being bleary-eyed early Sunday morning! Interesting that your version doesn't seem to set to `screen.$TERM` as suggested in the wiki and setting manually doesn't seem to help either...

**Reply by lingfish** [reply]
> Sorry yes I must have misread that - blame me still being bleary-eyed early Sunday morning!

All good, I'm doing this reply in the middle of Sunday night dinner 😆 

So, this from the screen manual (can't find a hotlink/anchor, search for section "16.1 Choosing the termcap entry for a window") explains a little bit at least, as to why I don't see a `screen.alacritty` (because that isn't in my TERMINFO as far as I can see):

> When screen tries to figure out a terminal name for itself, it first looks for an entry named screen.term, where term is the contents of your $TERM variable. If no such entry exists, screen tries ‘screen’ (or ‘screen-w’, if the terminal is wide (132 cols or more)). If even this entry cannot be found, ‘vt100’ is used as a substitute.

Where I'm lost though is this: why would specifically care about TERM?  What I mean is, if it's `screen`, or `vt52`, doesn't your app then do termcap/terminfo queries to determine features of the named terminal and proceed accordingly?  Why would you guys need to kinda "hard code" treatment for a `$TERM` of `screen`?

**Reply by TomJGooding** [reply]
I finally had some success with `screen` and `alacritty` with this setting:

```
~/.screenrc

term screen-256color
unsetenv COLORTERM
altscreen on
```

I think the problem is that `screen` doesn't support true colors, but  the `COLORTERM` was being set by alacritty.

~~However, there's still an issue where the app output isn't cleared on exit when run within `screen` (although I just tested with vim and its output isn't cleared on exit either)~~ Fixed the output not clearing by also setting `altscreen on`

Edit: Apparently true colors are supported in `screen v5.0.0` with `truecolor on`!

![SplashApp_2024-09-22T19_50_39_461703](https://github.com/user-attachments/assets/016d636c-d74b-485e-820e-3e3cde2dbd04)

**Reply by lingfish** [reply]
Ok, so here's what I discovered.

- The `TERM` setting, ending in `256color`, matters (eg. `screen-256color` or `xterm-...`) -- having just `screen` isn't enough
- The `COLORTERM=truecolor` matters -- no matter what else is set, this breaks it for me

So essentially, `unset COLORTERM` before anything else... then, with just `TERM=screen`, I get what perhaps looks like 8 bit colors (no surprise, halfway there).  If I run `TERM=xterm-256color pipenv run python -m textual`, I get a much better looking screen.  Same with the `screen-25...` variant.

**Reply by lingfish** [reply]
> Fixed the output not clearing by also setting `altscreen on`

Right, like the old-school xterm.  Funny, I dislike that behaviour :laughing: Can it not just place the cursor at the bottom of the terminal on exit?

**Reply by Yurii-huang** [reply]
You can't use a color scheme, and any program's screen will be displayed in black and white. The premise of using color should first confirm whether the color screen configuration is supported, I have a good scenario for using Tmux, obviously Tmux supports xterm-256color

**Reply by lingfish** [reply]
> You can't use a color scheme, and any program's screen will be displayed in black and white. The premise of using color should first confirm whether the color screen configuration is supported, I have a good scenario for using Tmux, obviously Tmux supports xterm-256color

I'm not sure what you point is here.
